### ISSUE TITLE
extern struct recognized differently from code translated from C

### ISSUE THREAD

**Issue by McSinyx** [ziglang/zig#9698]
I am converting a codebase from C++ to Zig.  So far, I switched the entry point to Zig and recreate the C++ objects in Zig.  The rest in the middle are still (extern "C") C++, i.e. we have a Zig sandwich here.

In order for the C++ code to use the Zig struct, there must be a header file and unfortunately it is also translated to Zig.  Thus, this shared module is recognized both as originally declared and `.cimport[…].struct_[…]` and the (stage 1) compiler yields the `error: expected type […], found […]`.  I made a [minimal reproducer here](https://gist.github.com/McSinyx/2626241649133de7170139061702305f) hoping it'll help debugging.

I also noticed that stage 1 compiler has disabled C header emission (GH-6753), was this one of the reasons?

**Comment by Vexu** [comment]
All struct types are unique, you cannot mix the automatically translated struct types with handwritten definitions.

**Comment by McSinyx** [comment]
I suppose the right way to do it here will be to ***@***.***` the struct
from the C header?  What should happen when the header is emitted
by `zig build`?  What I wrote by hand was identical to what -femit-h
gives, except for the C++ compatibility.

**Comment by McSinyx** [comment]
Wow using emails with GitHub is awful, this is the original message:

> I suppose the right way to do it here will be to `@cImport` the struct
> from the C header?  What should happen when the header is emitted
> by `zig build`?  What I wrote by hand was identical to what -femit-h
> gives, except for the C++ compatibility.

**Comment by Vexu** [comment]
> I suppose the right way to do it here will be to `@cImport` the struct
> from the C header?

Either that or add casts where needed.

>  What should happen when the header is emitted
> by `zig build`? What I wrote by hand was identical to what -femit-h
> gives, except for the C++ compatibility.

Unless there is a dependency loop, this should be pretty easy to do by chaining a `EmitHStep` (once such exists), to a `TranslateCStep` and then using that as a package.

**Comment by McSinyx** [comment]
Wouldn't the translated struct from emitted header still be different from the original one?  I don't see how that will make the `useFoo(makeFoo())` call in Zig legal.

**Comment by babelchips** [comment]
I agree. I'm inclined to think we'll have different structs in play even when headers can be emitted.

For example when writing a combination of Zig and C code where you wish to share a Zig defined structure:
1. Emit a new Zig packed struct (assuming the emit feature exists)
2. Include the header in your C code
3. From C, call into an exported Zig function, passing the struct (or a reference to it) as an argument
4. Surely the compilation will fail with `error: expected type […], found […]` ... ?

You couldn't even use cImport with cInclude of the exported header back in Zig because the header file would not exist initially (or be out of date).  Not that you would want to use imported versions of your own packed struct in the Zig code.

So, outside of using void* and casting either side, what is the ultimate solution here?

**Comment by SpexGuy** [comment]
You should define your structs in only one place.  If you define them on the C side, you can cImport that in exactly one place, and everything will work.  If you define them on the Zig side, you will also need to write your own extern function definitions, or do a post-process the generated .zig file from translate-c to fix up references.

**Comment by McSinyx** [comment]
> You should define your structs in only one place.

Agreed, but this doesn't account for emitted structs.  The only reason the struct was hand-written in the example was because emit-h has been disabled for stage 1 since 0.7.  I think that seamless interoperation with C is one of the goals of Zig, and getting structures *defined in Zig* to work in C should be a supported use case, as it's crucial for writing C ABI libraries *in Zig*.

**Comment by andrewrk** [comment]
> getting structures _defined in Zig_ to work in C should be a supported use case, as it's crucial for writing C ABI libraries _in Zig_.

Completely agreed. The plan is to resurrect the emit-h feature in the self-hosted compiler, which is under active development.

**Comment by Renha** [comment]
The "Add a Zig compilation unit to C/C++ projects" point in "⚡ Maintain it with Zig" section of main advertised language features should be removed until the header generation is back in a working state

**Comment by mlugg** [comment]
@Renha, why? That feature works entirely as described - header generation is a completely separate issue.

**Comment by Renha** [comment]
So you could add Zig unit to the project, but not use it, and that was the intention?

**Comment by abhinav** [comment]
@Renha emit-h doesn't affect what's currently listed in that section on the website.

> Use Zig as a zero-dependency, drop-in C/C++ compiler that supports cross-compilation out-of-the-box.

You can use `zig cc` and `zig c++` as drop-in C/C++ compilers.

> Leverage `zig build` to create a consistent development environment across all platforms.

You can write a `build.zig` that builds an existing C/C++ project without any changes to the code itself.
You can then run `zig build` instead of `make` or `cmake` or whatever else the project uses.

> Add a Zig compilation unit to C/C++ projects

You can expose a C ABI from code written in Zig, and link that against a C/C++ project.

None of these features need emit-h, but the last one becomes more convenient with it.
Right now, it's best to hand-write a header file for the C ABI you expose from your Zig library.
Fixing emit-h would generate the header file for you automatically.

**Comment by Renha** [comment]
Thank you, sorry if my tone sounded too harsh, I hope things will improve soon so that I would be able to use Zig in my project.

**Comment by mcharytoniuk** [comment]
As a workaround, you can usually get good results if you ask an LLM to generate C headers for you from the Zig code. They are good at converting between formats.

**Comment by saltzm** [comment]
@abhinav Do you know of any public examples of handwritten headers for zig libraries that I could look at?

**Comment by abhinav** [comment]
@saltzm Yeah, sure! Off the top of my head:

- libxev: https://github.com/mitchellh/libxev/blob/main/include/xev.h
- tigerbeetle: https://github.com/tigerbeetle/tigerbeetle/tree/main/src/clients/c (technically, they have [their own code generator](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/clients/c/tb_client_header.zig) for the C header based on the Zig types—it doesn't use emit-h)

If I had to guess, most projects using Zig 0.11 or newer, and exposing a C API would be hand-writing them right now.

**Comment by Des-Nerger** [comment]
As of `0.13.0-dev.344+b2588de6c`, a C header is generable through `$ zig test -femit-h src/root.zig`, but what about making the same through `build.zig`? I tried a solution suggested on Discord:
```
const install_lib = b.addInstallArtifact(lib, .{ .h_dir = .{ .override = .header } });
install_lib.emitted_h = lib.getEmittedH();
b.getInstallStep().dependOn(&install_lib.step);
```
, but got this: [EmittedH_0.13.0-dev.344+b2588de6c.log](https://github.com/ziglang/zig/files/15506018/EmittedH_0.13.0-dev.344%2Bb2588de6c.log). Is it a bug?

**Comment by ion232** [comment]
For what it's worth, I've been able to find a method that works for me as part of `build.zig`. It's a bit jank in that the build executable ends up picking up the exports, but it gets the job done. I've also set it up so that I can change the casing depending on whether I'm exporting to C or WASM. My project is still work in progress, but this method works for me so far. Unfortunately, I can't use externs in these exported functions because they get included in the build executable and the lack of symbols causes the build of the build executable to fail. I decided to just not use them to get around this, which is fine for my use case.

Here is where I generate the header in `build.zig`.
https://github.com/ion232/reticulum-zig/blob/main/build.zig#L59

And this is how I define the generate header function.
https://github.com/ion232/reticulum-zig/blob/main/core/exports.zig#L124

EDIT: The above comment is out of date, as I've improved upon this method by using the zig parser to generate bindings instead and this way avoids the extern issue.

See https://github.com/ion232/reticulum-zig/blob/main/core/bindings.zig
